agrest-cayenne/src/test/java/io/agrest/cayenne/processor/CayenneQueryAssemblerTest.java

```diff
@@ -103,6 +107,46 @@ public void createRootQuery_Qualifier() {
         assertEquals(E1.NAME.eq("X").andExp(E1.NAME.in("a", "b")), q2.getWhere());
     }
 
+    @Test
+    public void createRootQuery_Qualifier_Exists() {
+        RootResourceEntity<E3> entity = getResourceEntity(E3.class);
+
+        SelectContext<E3> c = new SelectContext<>(
+                E3.class,
+                new RequestSchema(mock(AgSchema.class)),
+                mock(AgRequestBuilder.class),
+                PathChecker.ofDefault(),
+                mock(Injector.class));
+
+        c.setEntity(entity);
+
+        entity.andExp(Exp.parse("exists e2"));
+        ObjectSelect<E3> q1 = queryAssembler.createRootQuery(c);
+        Expression expectedQ1 = ExpressionFactory.exists(ObjectSelect.query(E3.class).column(E3.E2));
+        assertEquals(expectedQ1, q1.getWhere());
+
+        entity.andExp(Exp.parse("not exists e5"));
+        ObjectSelect<E3> q2 = queryAssembler.createRootQuery(c);
+        assertEquals(expectedQ1.andExp(ExpressionFactory.notExists(ObjectSelect.query(E3.class).column(E3.E5))), q2.getWhere());
+    }
+
+    @Test
+    public void createRootQuery_Qualifier_Exists_Invalid_Condition() {
+        RootResourceEntity<E3> entity = getResourceEntity(E3.class);
+
+        SelectContext<E3> c = new SelectContext<>(
+                E3.class,
+                new RequestSchema(mock(AgSchema.class)),
+                mock(AgRequestBuilder.class),
+                PathChecker.ofDefault(),
+                mock(Injector.class));
+
+        c.setEntity(entity);
+
+        entity.andExp(Exp.exists("name = 'test1'"));
+        assertThrows(ExpressionException.class, () -> queryAssembler.createRootQuery(c));
+    }
```

Comment:
I think we could try to catch this case when we are assembling Cayenne Expression in the `CayenneExpressionVisitor`. For instance in the linking child to the parent method, or some other place where it will look good.